src/pretix/base/invoice.py

```diff
@@ -62,6 +62,63 @@
 logger = logging.getLogger(__name__)
 
 
+def addon_aware_groupby(iterable, key, is_addon):
+    """
+    We use groupby() to visually group identical lines on an invoice. For example, instead of
+
+    Product 1       5.00 EUR
+    Product 1       5.00 EUR
+    Product 1       5.00 EUR
+    Product 2       7.00 EUR
+
+    We want to print
+
+    3x Product 1    5.00 EUR = 15.00 EUR
+    Product 2       7.00 EUR
+
+    However, this fails for setups with addon-products since groupby() only groups consecutive
+    lines with the same identity. So in
+
+    Product 1       5.00 EUR
+    + Addon 1       2.00 EUR
+    Product 1       5.00 EUR
+    + Addon 1       2.00 EUR
+    Product 1       5.00 EUR
+    + Addon 2       3.00 EUR
+
+    There is no consecutive repetition of the same entity. This function provides a specialised groupby which
+    understands the product/addon relationship and packs groups of these addons together if they are, in fact,
+    identical groups:
+
+    2x Product 1    5.00 EUR = 10.00 EUR
+    + 2x Addon 1    2.00 EUR =  4.00 EUR
+    Product 1       5.00 EUR
+    + Addon 2       3.00 EUR
+    """
+    packed_groups = []
+
+    for i in iterable:
+        if is_addon(i):
+            packed_groups[-1][1].append(i)
+        else:
+            packed_groups.append((i, []))
+    # Each packed_groups element contains (parent product, list of addon products)
+
+    def _reorder(packed_groups):
+        # Emit the products as individual products again, reordered by "all parent products, then all addon products"
+        # within each group.
+        for group_grouper, grouped in groupby(packed_groups, key=lambda g: key(g[0]) + tuple(key(a) for a in g[1])):
+            grouped = list(grouped)
+            for grp in grouped:
+                yield grp[0]
+            for i in range(max(len(grp[1]) for grp in grouped)):
```

Comment:
Yikes. This line was hard to understand. Just as an exercise I thought about how I would do it: using `zip_longest` and checking for None instead of the index :3

Reply:
The addon lists (contents of `grouped[n][1]`) are always the same length, by definition (because we use them as grouping key above). Therefore we can just zip them and don't need to filter at all:
```
            parents, addon_groups = zip(*grouped)
            yield from parents
            for addons in zip(*addon_groups):
                yield from addons
```